### PR TITLE
[JBPM-9886] Task suspension and automatic reactivation after defined time period in jBPM

### PR DESCRIPTION
**Pull request by elguardian**
Jira: https://issues.redhat.com/browse/JBPM-9886

### CODE REVIEW

**Comment by mbiarnes** [comment]
Jenkins retest this

**Comment by mbiarnes** [comment]
Jenkins retest this

**Comment by mbiarnes** [comment]
Jenkins retest this

**Comment by mbiarnes** [comment]
Jenkins test

**Comment by Ginxo** [comment]
Jenkins retest this

**Comment by mareknovotny** [comment]
jenkins retest this please

File: jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/HumanTaskNodeInstance.java [hunk 1]
```diff
@@ -18,21 +18,43 @@
 
 import org.drools.core.process.instance.WorkItem;
 import org.jbpm.process.core.context.swimlane.SwimlaneContext;
+import org.jbpm.process.instance.InternalProcessRuntime;
 import org.jbpm.process.instance.context.swimlane.SwimlaneContextInstance;
+import org.jbpm.process.instance.timer.TimerInstance;
+import org.jbpm.process.instance.timer.TimerManager;
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/HumanTaskNodeInstance.java:24]: removed

File: jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/HumanTaskNodeInstance.java [hunk 2]
```diff
@@ -18,21 +18,43 @@
 
 import org.drools.core.process.instance.WorkItem;
 import org.jbpm.process.core.context.swimlane.SwimlaneContext;
+import org.jbpm.process.instance.InternalProcessRuntime;
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/HumanTaskNodeInstance.java:21]: removed

File: jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/HumanTaskNodeInstance.java [hunk 3]
```diff
@@ -93,4 +115,94 @@ public void triggerCompleted(WorkItem workItem) {
...
+
+    private boolean ownsWorkItem(Object event) {
+        if(!(event instanceof WorkItem)) {
+            return false;
+        }
+        WorkItem workItem = (WorkItem) event;
+        return (getWorkItemId() == workItem.getId() || (getWorkItemId() == -1 && getWorkItem().getId() == workItem.getId())) ;
+    }
+
+    private void humanTaskActivated(WorkItem event) {
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/HumanTaskNodeInstance.java:160]: removed

File: jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/HumanTaskNodeInstance.java [hunk 4]
```diff
@@ -93,4 +115,94 @@ public void triggerCompleted(WorkItem workItem) {
...
+        if(this.suspendUntilTimerId == -1) {
+            return;
+        }
+        ((WorkflowProcessInstanceImpl) getProcessInstance()).cancelTimer(suspendUntilTimerId);
+        suspendUntilTimerId = -1;
+    }
+
+    private void humanTaskSuspended(WorkItem workItem) {
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/HumanTaskNodeInstance.java:168]: removed

File: jbpm-human-task/jbpm-human-task-core/src/main/java/org/jbpm/services/task/events/WorkflowBridgeTaskLifeCycleEventListener.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    @Override
+    public void afterTaskSuspendedEvent(TaskEvent event) {
+        executeWork(event.getTask(), "humanTaskNodeInstance:suspended");
+    }
+
+    @Override
+    public void afterTaskActivatedEvent(TaskEvent event) {
+        executeWork(event.getTask(), "humanTaskNodeInstance:activated");
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-human-task/jbpm-human-task-core/src/main/java/org/jbpm/services/task/events/WorkflowBridgeTaskLifeCycleEventListener.java:43]: done

File: jbpm-human-task/jbpm-human-task-core/src/main/java/org/jbpm/services/task/events/WorkflowBridgeTaskLifeCycleEventListener.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+    private RuntimeManager runtimeManager;
+
+    public WorkflowBridgeTaskLifeCycleEventListener(RuntimeManager runtimeManager) {
+        this.runtimeManager = runtimeManager;
+    }
+
+    @Override
+    public void afterTaskSuspendedEvent(TaskEvent event) {
+        executeWork(event.getTask(), "humanTaskNodeInstance:suspended");
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-human-task/jbpm-human-task-core/src/main/java/org/jbpm/services/task/events/WorkflowBridgeTaskLifeCycleEventListener.java:38]: done

File: jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/HumanTaskNodeInstance.java [hunk 7]
```diff
@@ -93,4 +115,94 @@ public void triggerCompleted(WorkItem workItem) {
...
+        }
+
+        super.signalEvent(type, event);
+
+        if (!ownsWorkItem(event)) {
+            return;
+        }
+        if ("humanTaskNodeInstance:suspended".equals(type)) {
+            humanTaskSuspended((WorkItem) event);
+        } else if ("humanTaskNodeInstance:activated".equals(type)) {
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/HumanTaskNodeInstance.java:146]: done

File: jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/HumanTaskNodeInstance.java [hunk 8]
```diff
@@ -93,4 +115,94 @@ public void triggerCompleted(WorkItem workItem) {
...
+                return;
+            }
+        }
+
+        super.signalEvent(type, event);
+
+        if (!ownsWorkItem(event)) {
+            return;
+        }
+        if ("humanTaskNodeInstance:suspended".equals(type)) {
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/HumanTaskNodeInstance.java:144]: done

File: jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/HumanTaskNodeInstance.java [hunk 9]
```diff
@@ -93,4 +115,94 @@ public void triggerCompleted(WorkItem workItem) {
...
+
+        if("timerTriggered".equals(type)) {
+            TimerInstance timerInstance = (TimerInstance) event;
+            if(timerInstance.getId() == suspendUntilTimerId) {
+                // if there is suspended until we need to activate the task with the same user id was the owner.
+                RuntimeManager runtimeManager = RuntimeManagerRegistry.get().getManager(getProcessInstance().getDeploymentId());
+                RuntimeEngine engine = runtimeManager.getRuntimeEngine(ProcessInstanceIdContext.get(timerInstance.getProcessInstanceId()));
+                try {
+                    Task task = engine.getTaskService().getTaskByWorkItemId(getWorkItemId());
+                    engine.getTaskService().resume(task.getId(), task.getTaskData().getActualOwner().getId());
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/HumanTaskNodeInstance.java:130]: wrong user... it should be the user performing the operation.
thanks for pointing out.

File: jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/HumanTaskNodeInstance.java [hunk 10]
```diff
@@ -93,4 +117,96 @@ public void triggerCompleted(WorkItem workItem) {
...
+        if("timerTriggered".equals(type)) {
+            TimerInstance timerInstance = (TimerInstance) event;
+            if(timerInstance.getId() == suspendUntilTimerId) {
+                // if there is suspended until we need to activate the task with the same user id was the owner.
+                RuntimeManager runtimeManager = RuntimeManagerRegistry.get().getManager(getProcessInstance().getDeploymentId());
+                RuntimeEngine engine = runtimeManager.getRuntimeEngine(ProcessInstanceIdContext.get(timerInstance.getProcessInstanceId()));
+                try {
+                    Task task = engine.getTaskService().getTaskByWorkItemId(getWorkItemId());
+                    // setToPreviousStatus = true of resume therefore it will recover the old status
+                    // the operation is auto therefore use is administrator
```
Comment (gmunozfe) [inline jbpm-flow/src/main/java/org/jbpm/workflow/instance/node/HumanTaskNodeInstance.java:133]: typo "user"

**Review by gmunozfe — APPROVED** [review]
Looks good to me, awesome work @elguardian !
Rerunning the CI to get green

**Comment by gmunozfe** [comment]
jenkins retest this

**Comment by gmunozfe** [comment]
jenkins retest this

**Comment by gmunozfe** [comment]
jenkins retest this

**Comment by elguardian** [comment]
failure not related

**Pull request merged**